Attempt at implementing clients for Pub/Sub

dhermes opened the pull request:
- It doesn't seem that in pub/sub there was ever a use of `_LocalStack` so the worries about the welded on client and batching don't seem to manifest in the `pubsub` subpackage.
- I tried to make the commits as small and reviewable as possible.
- I'm happy to split this up into many PRs to make review / commenting easier.
- The `Using client on all Topic methods.` and `Using client on all Subscription methods.` commits are a bit longer than the rest. I'm also happy to break them up method-by-method if it makes review easier.
- I didn't look too closely, but there may be some superfluous connection vs. client tests that exist as artifacts of the old / transition states of the code.

jgeewax commented on docs/pubsub-usage.rst:
```diff
@@ -20,24 +32,27 @@ Create a new topic for the default project:
-   >>> topic = Topic('topic_name')
+   >>> from gcloud import pubsub
+   >>> client = pubsub.Client()
+   >>> topic = client.topic('topic_name')
```
I know it breaks the "one and only one way to do something" but is it crazy to offer the `create_topic` API method in addition?

I'm thinking...

```python
class Client(...):
  def create_topic(self, name):
    topic = self.topic(name)
    topic.create()
    return topic
```

This was partly due to the (what I believe correct) stance that `.create()` shouldn't return the topic, however if I'm in a REPL, I can see `client.create_topic()` being a very nice-to-have thing.

dhermes replied:
I think there are two chief issues with doing this

- The name `create_topic` is ambiguous to new users. Does it mean create a `Topic` instance or actual `insert` one?
- We'd need to offer `Client.topic` and `Client.create_topic` side-by-side since users would need to construct pre-existing topics without sending `insert`.

jgeewax replied:
> The name create_topic is ambiguous to new users. Does it mean create a Topic instance or actual insert one?

That's a good point. I think we should adopt the attitude that `client.<verb>_<noun>()` is an API request. So `create_topic` would hit the API trying to 'insert' the topic (using Google's API lingo). `client.<noun>` is nothing more than a factory for the noun which won't fire an API request but gives you back an instance of the <noun>.

> We'd need to offer Client.topic and Client.create_topic side-by-side since users would need to construct pre-existing topics without sending insert.

Totally agree -- I think we had this debate already and I'm totally sold on the need for `.topic()` as the sole way to get a hold of a topic object, without hitting the API (as the topic literally has *zero* extra metadata).

For other APIs, where the object itself holds lots of (potentially important) metadata, I think it might be worthwhile to provide a `get_<noun>()` to pull down that metadata, in addition to the `<noun>()` method which just gives you an object which may not be tied to the remote (authoritative) representation.

An example here might be `bucket`.

- `.bucket()` gives me a bucket object, sans remote data (no API request).
- `create_bucket()` explicitly creates the bucket in the remote service, throwing an error if I was unable to do so.
- `get_bucket()` would be an API request that gives me all the metadata about a bucket.

Note that the `<verb>_bucket()` methods are really just shortcuts so that I can have one-liners for the common things I want to do with a bucket. (ie, `create` == `bucket = client.bucket('name'); bucket.create(); return bucket;`)

In boto for S3, they have this `verify` or `check` parameter on the bucket constructor, so the comparisons look like:

Action|boto|gcloud (suggestion)
---|---|---
Get a bucket, no API request|`connection.get_bucket(validate=False)`|`client.bucket()`
Get a bucket + metadata|`connection.get_bucket()`|`client.get_bucket()`

I personally don't love the boto method, and would much rather the "gcloud (suggestion)" option.

dhermes replied:
Yeah we should port that large comment to an issue.

jgeewax replied:
OK I'm on it.

jgeewax replied:
#911

jgeewax commented on docs/pubsub-usage.rst:
```diff
@@ -20,24 +32,27 @@ Create a new topic for the default project:
-   >>> topic = Topic('topic_name', project='my.project')
+   >>> from gcloud import pubsub
+   >>> client = pubsub.Client()
+   >>> topic = Topic('topic_name', client=client)
```
I 100% agree that this should exist (I absolutely see the uses for it), but should we explain that in the full usage here? I would more expect that we'd recommend something like...

```python
client = pubsub.Client()
client2 = pubsub.Client.with_service_account_json('key.json')

topic = client.topic('name')
explicit_topic = client2.topic('name')
```

If someone wants to look at the method and see "oh cool, I can pass in an explicit client", that's fine. But I'd rather we recommend that you create a separate client and then use that for your interaction -- no?

dhermes replied:
Yeah I was just trying to map to what was previously there. I'm happy to nuke the section.

dhermes replied:
WDYT? Shall I nuke it?

jgeewax replied:
I think nuking the example is good (not nuking the parameter though, just the usage example of it)

dhermes replied:
Gone and pushed over original commit.

jgeewax commented on docs/pubsub-usage.rst:
```diff
@@ -20,24 +32,27 @@ Create a new topic for the default project:
    >>> topic.create()  # API request
 
-Check for the existance of a topic:
+Check for the existence of a topic:
```
:+1:

jgeewax commented on docs/pubsub-usage.rst:
```diff
@@ -20,24 +32,27 @@ Create a new topic for the default project:
-   >>> topic = Topic('topic_name')
+   >>> from gcloud import pubsub
+   >>> client = pubsub.Client()
+   >>> topic = client.topic('topic_name')
```
Same note as above -- would it be useful to have a `get_topic()` method (or similar) that does this API request in one go?

jgeewax commented on docs/pubsub-usage.rst:
```diff
@@ -45,17 +60,9 @@ List topics for the default project:
-   >>> topics, next_page_token = list_topics(project='my.project')  # API request
+   >>> from gcloud import pubsub
+   >>> client = pubsub.Client()
+   >>> topics, next_page_token = client.list_topics()  # API request
```
Longer discussion over on #895 -- I think we decided last time to not discuss that as part of the PR, but wanted your opinion regardless.

dhermes replied:
Commented there

jgeewax commented on docs/pubsub-usage.rst:
```diff
@@ -116,41 +127,42 @@ Create a new pull subscription for a topic:
+   >>> from gcloud import pubsub
+   >>> client = pubsub.Client()
+   >>> topic = client.topic('topic_name')
+   >>> subscription = pubsub.Subscription('subscription_name', topic)
```
Wouldn't we recommend that you get a subscription from a topic object?

```python
client = pubsub.Client()
topic = client.topic('topic-name')
sub = topic.subscription('subscription-name')
```

?

Also seems weird that I get a topic from a client, but a subscription directly from the module, if anything, wouldn't it be:

```python
client = pubsub.Client()
topic = client.topic('name')
sub = client.subscription('name', topic=topic)
sub_alternate = topic.subscription('name')  # Or get_subscription?
```

(I'm not saying get rid of the `topic` argument in the subscription constructor, but it seems like we'd want to say "the recommended way" is to get a topic, then pull a subscription from the topic.)

dhermes replied:
Yeah I felt that way when writing this. The only goal of this PR (with a very high number of commits) is to implement the client pattern.

jgeewax replied:
SGTM

jgeewax replied:
Added #912

jgeewax commented on docs/pubsub-usage.rst:
```diff
@@ -213,13 +228,13 @@ Fetch pending messages for a pull subscription:
+   ...     batch.publish('this is the first message_payload')
+   ...     batch.publish('this is the second message_payload',
    ...                   attr1='value1', attr2='value2')
    >>> received = subscription.pull()  # API request
```
I think this might be a bigger thing (aka, I should create a separate issue), but wouldn't we want to get a list (or iterator) of objects back, which has all the stuff we're interested in, rather than a tuple of the `(ackId, message)` ?

That is, I think it might be nicer to write:

```python
messages = subscription.pull()
for message in messages:
  print message.getAckId()  # ReceivedMessage.ackId (string)
  print message['attr1']  # PubsubMessage.attributes (dict)
  print message.getData()  # PubsubMessage.data (bytes)
  print message.getId()  # corresponds to PubsubMessage.messageId (string)
```

dhermes replied:
@tseaver WDYT?

PS @jgeewax we'd use PEP8 method names `get_ack_id` etc. (context switching from `node` to `python` can be difficult, I know)

jgeewax replied:
Ha - yes... you nailed it right on the head. Updating my comment to be PEP8 and not Node/Java

jgeewax replied:
I'm moving that over to a new issue. #913

jgeewax commented on gcloud/pubsub/client.py:
```diff
@@ -0,0 +1,211 @@
+        self.connection = Connection(credentials=credentials, http=http)
+
+    @classmethod
+    def with_service_account_json(cls, json_credentials_path, project=None):
```
Is `with` the right verb? I'd expect `from` but I could be wrong here.

dhermes replied:
We used `from_` in `Connection` but it doesn't really work there either. If the factory were directly creating a credentials object, then `from_` would make sense, but this one has some other parts in addition to the credentials piece, which is why I went with "with".

@tseaver typically has better insight on the use of language than I do

jgeewax replied:
Got it - that makes sense. SGTM.

tseaver replied:
I can live with either `with` or `from`.  Given that apps will call it as `Client.with_`, I think I have a preference for `from`.

dhermes replied:
SGTM. Prepositions gonna preposition

jgeewax commented on docs/pubsub-usage.rst:
```diff
@@ -1,16 +1,28 @@
 
-  - credentials (derived from GAE / GCE environ if present).
+- :class:`Client <gcloud.pubsub.client.Client>` objects hold both a ``project``
+  and an authenticated connection
```
... "to the PubSub service."

dhermes replied:
Done.

jgeewax commented on gcloud/pubsub/client.py:
```diff
@@ -0,0 +1,211 @@
+    """
+    def __init__(self, project=None, credentials=None, http=None):
+        if project is None:
+            project = _get_production_project()
```
Should we follow the `_require_X` pattern we had elsewhere? Should I open another issue for that?

dhermes replied:
It's not worth it here. The place this code comes from is a close cousin of `_implicit_environ`, all of which will get blown away when implementing clients.

Also note there is a local version of `_require` in each of `Topic` and `Subscription`.

jgeewax replied:
OK. SGTM.

jgeewax commented:
Awesome @dhermes . This LGTM but I think it may be worthwhile to have @tseaver take a look too.....

dhermes commented:
@tseaver You all done with your review?

tseaver commented:
Yup

dhermes commented:
@tseaver I addressed all feedback in a single commit (not my typical behavior, but figured that was the sanest thing to do given the size of this PR).

As mentioned in #912 this will fail the linter due to a cyclic import, but wanted to get it out for you to see anyhow.

jgeewax commented:
Status update on this one? Really liked the idea of getting this merged...

dhermes commented:
@jgeewax Sorry this fell by the wayside because 1f80e368d45da5036e4f753d25f48a47fd758160 couldn't successfully implement the `Topic.subscription` factory without a cyclic import.

Meant to discuss with @tseaver but didn't a chance to before he headed to Japan.

I can jettison that feature, get this PR in and then discuss in a new bug (and keep the code around in a new PR).

WDYT?

jgeewax commented:
SGTM

dhermes commented:
Filed #929 to discuss the `Topic <-> Subscription` cyclic issue and will put those changes in a new PR.
